### ISSUE TITLE
No plugin mechanism for runnables

### ISSUE THREAD

**Issue by bonzini** [avocado-framework/avocado#3395]
Resolvers can create arbitrary kinds of runnables, but only those included in RUNNABLE_KIND_CAPABLE can then actually be used. Likewise you cannot do `python3 -m avocado.core.nrunner runnable-run -k glib -u ./some-test` because the `glib` kind is not included in the capabilities.

**Comment by bonzini** [comment]
In addition, when Resolvers for optional plugins were introduced in commit e0c2e2465, the corresponding BaseRunner subclasses were not added.

It's very unclear to me if the nextrunner concepts are even ready for developers to use them, so would new features be accepted for the old runner (for example different human logging options, even parallel tests, or adding TAP subtests to ResultEvents)

**Comment by clebergnu** [comment]
Hi @bonzini, short answer is that you've found out *how* experimental resolvers and nrunner currently are.  In short, they're are *not* ready for developers that expect a stable interface and complete experience.

On Avocado 73.0, though, an interface for different runners (not be be confused with the nrunner concept of runners) was introduced, and you can see the current implementations with:

```
$ avocado plugins
...
Plugins that run test suites on a job (runners): 
docker Runs on a Docker (or compatible) container
remote Runs on a remote machine using SSH
runner The conventional test runner
vm     Runs on a VM using SSH
```

I'm now working on a "nrunner" plugin for that interface, which means that all other features should transparently work (like the `ResultEvents` interface and plugins).  As a general rule, the nrunner should *not* block any work on the current runner.  The real question is how much effort one should put into the runner feature (that is already covered by the nrunner, such as parallel execution).  In the case of `ResultEvents`, I don't see any overlap though.

Now, back to the `RUNNABLE_KIND_CAPABLE` question, there are two ways that the `nrun` looks for runner implementations:

1) Python modules with name matching `avocado.core.nrunner_%(kind)s`, which is planned to be used as an optimization when execution is local (see https://github.com/avocado-framework/avocado/blob/73.0/avocado/plugins/nrun.py#L97)
2) Any executable in the system PATH with name matching `avocado-runner-$(kind)s`, which is how runners written in any language can be supported (see https://github.com/avocado-framework/avocado/blob/73.0/avocado/plugins/nrun.py#L104).

At this point, I'm also considering reading information generated at run time by the runners to influence this selection, for example:

```
$ avocado-runner capabilities 
{"runnables": ["noop", "exec", "exec-test", "python-unittest"], "commands": ["capabilities", "runnable-run", "runnable-run-recipe", "task-run", "task-run-recipe", "status-server"]}
```

There are some downsides to this approach, so it's TBD.

**Comment by bonzini** [comment]
So if I changed the Runner in `avocado_glib` to a script installed to /usr/bin/avocado-runner-glib would that be accepted? I am okay with doing some of the work to make the nrunner more usable, because it's a better match for what I am working on.

(Where does nrunner do parallel execution BTW? I would like to add support for `make -j` so that multiple parallel invocations of Avocado do not create too many jobs).

**Comment by clebergnu** [comment]
A few extra extra points for the context of those decisions:

1) The current interface "Runner" interface[1]  and its implementations (`docker`, `remote`, `runner` `vm`) deal with a test suite (in truthfulness all tests of a job) at once
2) One goal of the nrunner is to able to run *individual* tests in fresh and isolated environments in parallel, with minimal dependencies, such as nothing but the `avocado-runner-$(kind)s` itself.

The second point suggests that relying on the previous setup of Python entrypoints on the machine that will run a test is a bad idea.  That's why the selection of a "nrunner runner" is done differently.  In factm the `avocado/core/nrunner.py` file should be transportable to *any* environment containing Python 3 and nothing else.

With regards to `python3 -m avocado.core.nrunner runnable-run -k glib -u ./some-test`, this could be done in a few different ways:
* `avocado nrun ./some-test` (assuming the resolver is properly resolving it as a `glib` test
* `avocado-runner-glib runnable-run -k glib -u ./some-test` (I'm aware of the redundancy in here)
* `avocado-runner runnable-run-recipe $path_to_recipe_file` assuming the `$path_to_recipe_file` contains the `kind`

[1] - https://github.com/avocado-framework/avocado/blob/73.0/avocado/core/plugin_interfaces.py#L290

**Comment by clebergnu** [comment]
> So if I changed the Runner in `avocado_glib` to a script installed to /usr/bin/avocado-runner-glib would that be accepted? I am okay with doing some of the work to make the nrunner more usable, because it's a better match for what I am working on.
> 

Yes, of course.  This is one example of how I did for `robot`:

https://github.com/avocado-framework/avocado/blob/73.0/optional_plugins/robot/avocado_robot/runner.py#L39
https://github.com/avocado-framework/avocado/blob/73.0/optional_plugins/robot/setup.py#L31

> (Where does nrunner do parallel execution BTW? I would like to add support for `make -j` so that multiple parallel invocations of Avocado do not create too many jobs).

It's very crude code, but here's the pointer:

https://github.com/avocado-framework/avocado/blob/73.0/avocado/plugins/nrun.py#L65

**Comment by bonzini** [comment]
Thanks, I will take a look and ask if I have questions.
